Port scripts to strict POSIX implementations

aaronhurt opened the pull request:
There is a new script in this PR `zfs-replicate-posix.sh` that adds POSIX shell support. This script should function via standard Bourne Shell (sh) and/or strict POSIX shells like Dash. This script is currently considered experimental. Feedback is welcome.

tschettervictor commented:
The script stops at these two lines

```
    srcSnaps=$(snapList "$src" "$srcHost" 1)
    dstSnaps=$(snapList "$dst" "$dstHost" 0)
```
When commenting them out, it proceeds, but it stops here.

aaronhurt commented:
> The script stops at these two lines 257-258
> 
> ```
>     srcSnaps=$(snapList "$src" "$srcHost" 1)
>     dstSnaps=$(snapList "$dst" "$dstHost" 0)
> ```
> 
> When commenting them out, it proceeds, but it stops here.

It won't work at all if those are commented out, hrm it works here.

tschettervictor commented:
> > The script stops at these two lines 257-258
> > ```
> >     srcSnaps=$(snapList "$src" "$srcHost" 1)
> >     dstSnaps=$(snapList "$dst" "$dstHost" 0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When commenting them out, it proceeds, but it stops here.
> 
> It won't work at all if those are commented out, hrm it works here. Maybe something specific about FreeBSD `sh` that `dash` does different. I'll check it on my FreeBSD host.
> 
> EDIT: You can help debug by editing the first line of the script from `set -e` to `set -ex` and you'll get verbose debugging output.

Yes I'm testing that way. The function works because it shows srcSnaps=' ' on the last line of the script.

aaronhurt commented:
This is working from my FreeBSD host with a basic test ...
```
[ahurt@schroeder ~/zfs-replicate]$ uname -a
FreeBSD schroeder.anbcs.com 14.1-RELEASE-p3 FreeBSD 14.1-RELEASE-p3 GENERIC amd64
[ahurt@schroeder ~/zfs-replicate]$ doas /bin/sh -c "LOG_BASE=./logs REPLICATE_SETS=zstore/temp:zstore/temp2 ./zfs-replicate-posix.sh"
Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: loading configuration from defaults and environmental settings.
Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: creating lockfile /tmp/.replicate.snapshot.lock
Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: checking dataset cmd=/sbin/zfs list -H -o name zstore/temp
Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: zstore/temp
Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: checking dataset cmd=/sbin/zfs list -H -o name zstore/temp2
Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: zstore/temp2
Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: found old snapshot zstore/temp@autorep-08292024_1724980392
Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: destroying snapshot cmd=/sbin/zfs destroy zstore/temp@autorep-08292024_1724980392
Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: creating snapshot cmd=/sbin/zfs snapshot zstore/temp@autorep-08292024_1724980551
Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: creating lockfile /tmp/.replicate.send.lock
Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: sending snapshot cmd=/sbin/zfs send -Rs -I zstore/temp@autorep-08292024_1724980442 zstore/temp@autorep-08292024_1724980551 | /sbin/zfs receive -vFd zstore/temp2
Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: WARNING: skipping dataset zstore/temp/foo and its children: snapshot autorep-08292024_1724980551 does not exist
Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: WARNING: skipping dataset zstore/temp/hello and its children: snapshot autorep-08292024_1724980551 does not exist
Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: attempting destroy zstore/temp2/temp@autorep-08292024_1724980392
Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: success
Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: receiving incremental stream of zstore/temp@autorep-08292024_1724980551 into zstore/temp2/temp@autorep-08292024_1724980551
Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: received 312B stream in 2.66 seconds (117B/sec)
Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: deleting lockfile /tmp/.replicate.send.lock
Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: deleting lockfile /tmp/.replicate.snapshot.lock
Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: pruning logs ./logs/autorep-08292024_1724979945.log
Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: success total sets 1 skipped 0
```

tschettervictor commented:
> This is working from my FreeBSD host with a basic test ...
> 
> ```
> [ahurt@schroeder ~/zfs-replicate]$ uname -a
> FreeBSD schroeder.anbcs.com 14.1-RELEASE-p3 FreeBSD 14.1-RELEASE-p3 GENERIC amd64
> [ahurt@schroeder ~/zfs-replicate]$ doas /bin/sh -c "LOG_BASE=./logs REPLICATE_SETS=zstore/temp:zstore/temp2 ./zfs-replicate-posix.sh"
> Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: loading configuration from defaults and environmental settings.
> Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: creating lockfile /tmp/.replicate.snapshot.lock
> Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: checking dataset cmd=/sbin/zfs list -H -o name zstore/temp
> Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: zstore/temp
> Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: checking dataset cmd=/sbin/zfs list -H -o name zstore/temp2
> Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: zstore/temp2
> Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: found old snapshot zstore/temp@autorep-08292024_1724980392
> Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: destroying snapshot cmd=/sbin/zfs destroy zstore/temp@autorep-08292024_1724980392
> Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: creating snapshot cmd=/sbin/zfs snapshot zstore/temp@autorep-08292024_1724980551
> Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: creating lockfile /tmp/.replicate.send.lock
> Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: sending snapshot cmd=/sbin/zfs send -Rs -I zstore/temp@autorep-08292024_1724980442 zstore/temp@autorep-08292024_1724980551 | /sbin/zfs receive -vFd zstore/temp2
> Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: WARNING: skipping dataset zstore/temp/foo and its children: snapshot autorep-08292024_1724980551 does not exist
> Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: WARNING: skipping dataset zstore/temp/hello and its children: snapshot autorep-08292024_1724980551 does not exist
> Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: attempting destroy zstore/temp2/temp@autorep-08292024_1724980392
> Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: success
> Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: receiving incremental stream of zstore/temp@autorep-08292024_1724980551 into zstore/temp2/temp@autorep-08292024_1724980551
> Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: received 312B stream in 2.66 seconds (117B/sec)
> Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: deleting lockfile /tmp/.replicate.send.lock
> Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: deleting lockfile /tmp/.replicate.snapshot.lock
> Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: pruning logs ./logs/autorep-08292024_1724979945.log
> Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: success total sets 1 skipped 0
> ```

Did you remove all snapshots from src and dst?

aaronhurt commented:
> > This is working from my FreeBSD host with a basic test ...
> > ```
> > [ahurt@schroeder ~/zfs-replicate]$ uname -a
> > FreeBSD schroeder.anbcs.com 14.1-RELEASE-p3 FreeBSD 14.1-RELEASE-p3 GENERIC amd64
> > [ahurt@schroeder ~/zfs-replicate]$ doas /bin/sh -c "LOG_BASE=./logs REPLICATE_SETS=zstore/temp:zstore/temp2 ./zfs-replicate-posix.sh"
> > Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: loading configuration from defaults and environmental settings.
> > Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: creating lockfile /tmp/.replicate.snapshot.lock
> > Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: checking dataset cmd=/sbin/zfs list -H -o name zstore/temp
> > Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: zstore/temp
> > Aug 29 20:15:51 zfs-replicate-posix.sh[48582]: checking dataset cmd=/sbin/zfs list -H -o name zstore/temp2
> > Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: zstore/temp2
> > Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: found old snapshot zstore/temp@autorep-08292024_1724980392
> > Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: destroying snapshot cmd=/sbin/zfs destroy zstore/temp@autorep-08292024_1724980392
> > Aug 29 20:15:52 zfs-replicate-posix.sh[48582]: creating snapshot cmd=/sbin/zfs snapshot zstore/temp@autorep-08292024_1724980551
> > Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: creating lockfile /tmp/.replicate.send.lock
> > Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: sending snapshot cmd=/sbin/zfs send -Rs -I zstore/temp@autorep-08292024_1724980442 zstore/temp@autorep-08292024_1724980551 | /sbin/zfs receive -vFd zstore/temp2
> > Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: WARNING: skipping dataset zstore/temp/foo and its children: snapshot autorep-08292024_1724980551 does not exist
> > Aug 29 20:15:53 zfs-replicate-posix.sh[48582]: WARNING: skipping dataset zstore/temp/hello and its children: snapshot autorep-08292024_1724980551 does not exist
> > Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: attempting destroy zstore/temp2/temp@autorep-08292024_1724980392
> > Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: success
> > Aug 29 20:15:55 zfs-replicate-posix.sh[48582]: receiving incremental stream of zstore/temp@autorep-08292024_1724980551 into zstore/temp2/temp@autorep-08292024_1724980551
> > Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: received 312B stream in 2.66 seconds (117B/sec)
> > Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: deleting lockfile /tmp/.replicate.send.lock
> > Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: deleting lockfile /tmp/.replicate.snapshot.lock
> > Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: pruning logs ./logs/autorep-08292024_1724979945.log
> > Aug 29 20:15:57 zfs-replicate-posix.sh[48582]: success total sets 1 skipped 0
> > ```
> 
> Did you remove all snapshots from src and dst?

No, I didn't but thinking about it that might make sense. In a POSIX world that could be an error - no count when it comes to the next few commands.

yonas reviewed (changes requested):
@aaronhurt This commit isn't signed. Can you please sign before merging?

yonas reviewed (changes requested):
`zfs-replicate-posix.sh` ran without error, but there's nothing in the destination dataset.

```sh
$ cat config.sh
REPLICATE_SETS="zroot/test:zroot/fun@localhost"
```

aaronhurt commented:
Fixed the issue with no snaps on either side, counting issue. The POSIX way of using `wc -l` will return a `1` for an empty string that's newline terminated.

tschettervictor commented:
> `zfs-replicate-posix.sh` ran without error, but there's nothing in the destination dataset.
> 
> ```shell
> $ cat config.sh
> REPLICATE_SETS="zroot/test:zroot/fun@localhost"
> ```

When replicating locally you dont really need the @localhost

yonas commented:
I recommend this small change:

```diff
f302e68 - (HEAD -> pr-41) Update example config script to use sh. (2 minutes ago) <Yonas> yonas.y@gmail.com SHA256:5ZC2NLyr/ExZaQKf7DPl2f6V65LX8roqeB01qtT4izE

diff --git a/config.sample.sh b/config.sample.sh
index afadb6d..a6342bd 100644
--- a/config.sample.sh
+++ b/config.sample.sh
@@ -1,4 +1,5 @@
-#!/usr/bin/env bash
+#!/usr/bin/sh
+
 ## zfs-replicate configuration file
 # shellcheck disable=SC2034
```

yonas commented:
> > `zfs-replicate-posix.sh` ran without error, but there's nothing in the destination dataset.
> > ```shell
> > $ cat config.sh
> > REPLICATE_SETS="zroot/test:zroot/fun@localhost"
> > ```
> 
> When replicating locally you dont really need the @localhost

Thanks :+1:

aaronhurt commented:
> I recommend this small change:
> 
> ```diff
> f302e68 - (HEAD -> pr-41) Update example config script to use sh. (2 minutes ago) <Yonas> yonas.y@gmail.com SHA256:5ZC2NLyr/ExZaQKf7DPl2f6V65LX8roqeB01qtT4izE
> 
> diff --git a/config.sample.sh b/config.sample.sh
> index afadb6d..a6342bd 100644
> --- a/config.sample.sh
> +++ b/config.sample.sh
> @@ -1,4 +1,5 @@
> -#!/usr/bin/env bash
> +#!/usr/bin/sh
> +
>  ## zfs-replicate configuration file
>  # shellcheck disable=SC2034
> ```
Yes, that should be changed if/when the Bash script is completely replaced with the POSIX version. Although, it probably doesn't hurt now since it's just an ENV file.

yonas commented:
> Fixed the issue with no snaps on either side. It was a counting issue. The POSIX way of using `wc -l` will return a `1` for an empty string that's newline terminated. The `grep` also needs to be fed to `|| true` or it will exit `1` when the string doesn't match.

Replication works for me now :tada:

aaronhurt commented:
It should still call be `#!/usr/bin/env sh` though. The `env` program is at `/usr/bin/env` 99.999% of the time. Whereas `sh` can be in `/bin/` or `/usr/bin` depending on the system.

yonas commented:
> It should still be `#!/usr/bin/env sh` though. The `env` program is at `/usr/bin/env` 99.999% of the time. Whereas `sh` can be in `/bin/` or `/usr/bin` depending on the system.

Yes, I agree. I just checked Ubuntu, and it was in `/usr/bin`.

aaronhurt commented:
Ubuntu `sh` is Dash I believe, and FreeBSD `sh` is POSIX Bourne shell.

aaronhurt commented:
I'm going to call it a night, but feel free to leave comments and I'll take a look tomorrow.

tschettervictor commented:
I will test with all combos and report back.

tschettervictor commented:
So. 
Local replication seems to work perfectly with RECURSE_CHILDREN=0/1 and ALLOW_RECONCILITATION=0/1 + nested/un-nested datasets
Remote replication also seems to work fine with the above listed combinations.

However, I ran into one problem. The way I'm running this test is by a REPLICATE_SETS="tank/local/src:tank/local/dst tank/src/1/2:rmttank/dst/1/2@1.2.3.4 tmttank/srt/1/2@1.2.3.4:tank/dst/1/2"
So I'm doing a local, a remote push, and a remote pull in the same go

For some reason it is kind of messing up the third set. I believe it might be due to the fact that we are doing a pull and push in the same run, but still, it needs to be fixed. When doing a pull separate from any other sets, it does work.

Also, when the above happens it doesn't error out with a fail. There are other cases where the script also did not error out, like the earlier issue we had, but that is probably a POSIX thing. It would be nice though to add some more error checking. The log did show that it didn't skip any datasets when using the above example, even though the third set failed at listing the dataset.

tschettervictor commented:
Looks like clearing the variable right before doing the "contains" function solves the issue. See my commit.
Ill do some more testing tomorrow evening.

yonas commented:
@tschettervictor Very nice.

I have just one last suggestion, since `sh` was [updated](https://svnweb.freebsd.org/base?view=revision&revision=345561):

```diff
diff --git a/zfs-replicate-posix.sh b/zfs-replicate-posix.sh
index 50af598..d2ddbcf 100755
--- a/zfs-replicate-posix.sh
+++ b/zfs-replicate-posix.sh
@@ -1,6 +1,6 @@
 #!/usr/bin/env sh
 ## zfs-replicate.sh
-set -e
+set -euo pipefail
 
 ############################################
 ##### warning gremlins live below here #####
 ```

yonas reviewed (approved):
Approved.

Last suggestion is adding `-o pipefail`:

https://github.com/aaronhurt/zfs-replicate/pull/41#issuecomment-2320846605

aaronhurt commented:
Thanks @yonas, it's  nice to see that `pipefail` is supported in FreeBSD Bourne shell now, but I'm sure it's not everywhere. I added a conditional check to test for `set -o pipefail` and add it if supported. I also added `-u` to fail on any unbound variables.

tschettervictor commented:
Do we want to test some more? From my end it looks good.

tschettervictor reviewed (approved):
I’ve tested all combinations using the new POSIX script. All have passed.

aaronhurt commented:
> I’ve tested all combinations using the new POSIX script. All have passed.

Nice, I'm working on new test cases, then I'll merge this into main.

tschettervictor commented:
> > I’ve tested all combinations using the new POSIX script. All have passed.
> 
> Nice, I'm working on new test cases, then I'll merge this into main. This script has gotten more activity in the past few days than it has in the last 7+ years 🤣

Indeed. I really appreciate it though, and all you’ve done in the last few days. This script could even serve as a remote control to replicate a remote server to another remote server.

tschettervictor commented:
Is it practical to keep the bash script along with the sh script?

aaronhurt commented:
> Is it practical to keep the bash script along with the sh script?

No, I don't want to support two different shell scripts that do the same thing.

tschettervictor commented:
Tiny note here. When doing a `zfs-replicate -s` it is not finding the log, even thought the LOGBASE is set. Could this be because it is not loading the variable before it is doing the "checkStatus" function?

aaronhurt commented:
That status bit should be fixed and with that That's it, I'm hands off. Let me know if you find anything else @tschettervictor or @yonas. Have a great weekend!

tschettervictor commented:
Hmmm… it’s not respecting the $ZFS at some points. When doing the send it respects it, but on the receive it uses the script default and not the one set inside the config file.

aaronhurt commented:
The tests all seem to work, and it sets a fake zfs to the script.

tschettervictor commented:
> The tests all seem to work, and it sets a fake zfs to the script.

It uses it everywhere but on one line.

Aug 30 17:06:15 zfs-replicate.sh[22277]: + /usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd rmtpool/rmtdst/1

aaronhurt commented:
> > The tests all seem to work, and it sets a fake zfs to the script.
> 
> It uses it everywhere but on one line.
> 
> Aug 30 17:06:15 zfs-replicate.sh[22277]: + /usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd rmtpool/rmtdst/1

I don’t see any blank receive calls.

tschettervictor commented:
> > > The tests all seem to work, and it sets a fake zfs to the script.
> > 
> > 
> > It uses it everywhere but on one line.
> > Aug 30 17:06:15 zfs-replicate.sh[22277]: + /usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd rmtpool/rmtdst/1
> 
> I don’t see any static receive calls. I had to force push to fix the conflicts. Make sure you are on the latest revision.

```
Aug 30 17:12:46 zfs-replicate.sh[25408]: + pipe='zfs receive -vFd'
Aug 30 17:12:46 zfs-replicate.sh[25408]: + [ -n 192.168.1.132 ]
Aug 30 17:12:46 zfs-replicate.sh[25408]: + printf '%s\n' '/usr/bin/ssh %HOST% /sbin/zfs receive -vFd'
Aug 30 17:12:46 zfs-replicate.sh[25408]: + sed s/%HOST%/192.168.1.132/g
Aug 30 17:12:46 zfs-replicate.sh[25408]: + pipe='/usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd'
Aug 30 17:12:46 zfs-replicate.sh[25408]: + pipe='/usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd RSCPOOL2/rmtdst/1'
```

aaronhurt commented:
Can you post your configuration, or at least the sets and pipes?

tschettervictor commented:
Everything is default and copied from the updated files except for

ZFS=zfs
REPLICATE_SETS="tank/src/1:tank/dst/1 tank/rmtsrc/1:POOL2/rmtdst/1@192.168.1.132 POOL2/rmtsrc/1@192.168.1.132:tank/rmtdst/1"
LOGBASE=“/var/log/zfs-replicate”

tschettervictor commented:
```
Aug 30 18:28:33 zfs-replicate.sh[47214]: + cmd='zfs send -Rs tank/rmtsrc/1@autorep-08302024_1725064111'
Aug 30 18:28:33 zfs-replicate.sh[47214]: + pipe='/sbin/zfs receive -vFd'
Aug 30 18:28:33 zfs-replicate.sh[47214]: + [ -n 192.168.1.132 ]
Aug 30 18:28:33 zfs-replicate.sh[47214]: + printf '%s\n' '/usr/bin/ssh %HOST% /sbin/zfs receive -vFd'
Aug 30 18:28:33 zfs-replicate.sh[47214]: + sed s/%HOST%/192.168.1.132/g
Aug 30 18:28:33 zfs-replicate.sh[47214]: + pipe='/usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd'
Aug 30 18:28:33 zfs-replicate.sh[47214]: + pipe='/usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd RSCPOOL2/rmtdst/1'
Aug 30 18:28:33 zfs-replicate.sh[47214]: + printf 'sending snapshot cmd=%s | %s\n' 'zfs send -Rs tank/rmtsrc/1@autorep-08302024_1725064111' '/usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd RSCPOOL2/rmtdst/1'
Aug 30 18:28:33 zfs-replicate.sh[47214]: sending snapshot cmd=zfs send -Rs tank/rmtsrc/1@autorep-08302024_1725064111 | /usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd RSCPOOL2/rmtdst/1
Aug 30 18:28:33 zfs-replicate.sh[47214]: + zfs send -Rs tank/rmtsrc/1@autorep-08302024_1725064111
Aug 30 18:28:33 zfs-replicate.sh[47214]: + /usr/bin/ssh 192.168.1.132 /sbin/zfs receive -vFd RSCPOOL2/rmtdst/1
Aug 30 18:28:33 zfs-replicate.sh[47214]: WARNING: skipping dataset tank/rmtsrc/1/2 and its children: snapshot autorep-08302024_1725064111 does not exist
Aug 30 18:28:34 zfs-replicate.sh[47214]: zsh:1: no such file or directory: /sbin/zfs

```
It appears to be an issue with the DEST_PIPE_WITH_HOST variable.

tschettervictor commented:
Maybe it would be possible to use the (which zfs) command as the actual command in such cases where the default zfs path is not the same on the remote and local sides. Right now it simply uses the script hosts default path.

aaronhurt commented:
> It appears to be an issue with the DEST_PIPE_WITH_HOST variable. When I uncomment that variable in the config file, everything works perfectly, but it it just used the default in the main script, then it fails for some reason.
> 
> EDIT Looks like the DEST_PIPE vars get set before the config file loads, that's why
> 
> EDIT2 Moving them (the two DEST_PIPE vars) down to after the config file is sourced solves this issue.

Do you have either `DEST_PIPE_*` defined in your config? The settings at the top of the file are just defaults. The config file will override them if they are present in the config.

aaronhurt commented:
```
DEST_PIPE_WITH_HOST="${DEST_PIPE_WITH_HOST:-"$SSH %HOST% $ZFS receive -vFd"}"
DEST_PIPE_WITHOUT_HOST="${DEST_PIPE_WITHOUT_HOST:-"$ZFS receive -vFd"}"
```

Those say set `DEST_PIPE_WITH_HOST` to the value of `DEST_PIPE_WITH_HOST` _if set_, else, set to `"$SSH %HOST% $ZFS receive -vFd"`, same with the other variable. They are also not `readonly` at this point, so if there is a value sourced in the config it will override those before being locked at the end of `loadConfig`.

tschettervictor commented:
> > It appears to be an issue with the DEST_PIPE_WITH_HOST variable. When I uncomment that variable in the config file, everything works perfectly, but it it just used the default in the main script, then it fails for some reason.
> > EDIT Looks like the DEST_PIPE vars get set before the config file loads, that's why
> > EDIT2 Moving them (the two DEST_PIPE vars) down to after the config file is sourced solves this issue.
> 
> Do you have either `DEST_PIPE_*` defined in your config? The settings at the top of the file are just defaults. The config file will override them if they are present in the config.

Nope. I never touch those in the config file.

tschettervictor commented:
> ```
> DEST_PIPE_WITH_HOST="${DEST_PIPE_WITH_HOST:-"$SSH %HOST% $ZFS receive -vFd"}"
> DEST_PIPE_WITHOUT_HOST="${DEST_PIPE_WITHOUT_HOST:-"$ZFS receive -vFd"}"
> ```
> 
> Those say set `DEST_PIPE_WITH_HOST` to the value of `DEST_PIPE_WITH_HOST` _if set_, else, set to `"$SSH %HOST% $ZFS receive -vFd"`, same with the other variable. They are also not `readonly` at this point, so if there is a value sourced in the config it will override those before being locked at the end of `loadConfig`.

I get the logic here, but it doesn’t seem to apply. DEST_PIPE seems to set itself at the top along with the ZFS, then later ZFS is overridden by the config file, but DEST_PIPE stays at the value it had at the beginning of the file, which was $(which zfs)

aaronhurt commented:
> > ```
> > DEST_PIPE_WITH_HOST="${DEST_PIPE_WITH_HOST:-"$SSH %HOST% $ZFS receive -vFd"}"
> > DEST_PIPE_WITHOUT_HOST="${DEST_PIPE_WITHOUT_HOST:-"$ZFS receive -vFd"}"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Those say set `DEST_PIPE_WITH_HOST` to the value of `DEST_PIPE_WITH_HOST` _if set_, else, set to `"$SSH %HOST% $ZFS receive -vFd"`, same with the other variable. They are also not `readonly` at this point, so if there is a value sourced in the config it will override those before being locked at the end of `loadConfig`.
> 
> I get the logic here, but it doesn’t seem to apply. DEST_PIPE seems to set itself at the top along with the ZFS, then later ZFS is overridden by the config file, but DEST_PIPE stays at the value it had at the beginning of the file, which was ZFS=$(which zfs)

Yes, that's correct .... the ZFS value will be substituted there. That's a fair callout. I added some tests for the config. Let me take that into account.

aaronhurt commented:
Done, and the test case covers just that example now.

tschettervictor commented:
Initial run passed. Testing commencing. Stand by…

tschettervictor commented:
Testing completed. Just for reference, here are my test runs again.


My REPLICATE_SETS include one local set, one remote push set, and one remote pull set.

- Run 3-4 times then delete destination snaps, then run again. This should pass.
- Run 3-4 times then delete source snaps, then run again. This should fail, unless ALLOW_RECONCILIATION=1
- Do the above with RECURSE_CHILDREN=0, then do it with RECURSE_CHILDREN=1
- Also do the above with ALLOW_RECONCILIATION=0 to make sure it fails when source snaps are missing, and passes even if they are missing and ALLOW_RECONCILIATION=1

Looks good from my end.

tschettervictor commented:
I do notice though (and this might be a good thing) that when the script encounters an error, it will exit even if it successfully completed replicating the first 2 pairs. This makes it so the end message is ERROR: etc...

But this has only happened when a bug has popped up. Otherwise it will show the "skipped" part at the end.

aaronhurt commented:
> I do notice though (and this might be a good thing) that when the script encounters an error, it will exit even if it successfully completed replicating the first 2 pairs. This makes it so the end message is ERROR: etc...
> 
> But this has only happened when a bug has popped up. Otherwise it will show the "skipped" part at the end.

Definitely document those in issues if you can reproduce them. They’ll make great test cases to add to further improve the script.

tschettervictor commented:
> > I do notice though (and this might be a good thing) that when the script encounters an error, it will exit even if it successfully completed replicating the first 2 pairs. This makes it so the end message is ERROR: etc...
> > But this has only happened when a bug has popped up. Otherwise it will show the "skipped" part at the end.
> 
> Definitely document those in issues if you can reproduce them. They’ll make great test cases to add to further improve the script.

https://github.com/aaronhurt/zfs-replicate/issues/43

aaronhurt commented:
If no-one is opposed I will merge this into the main branch this evening.

tschettervictor commented:
I’m good to merge. Test have all passed.

aaronhurt commented:
> I’m good to merge. Test have all passed.
> 
> I’ll try some more intentional errors.

Sounds good, let me know if you find anything. Otherwise I'll merge this tonight or tomorrow morning, but I want to give it a little time to bake.

tschettervictor commented:
But: when it skips a dataset that failed the “checkDataset “ function, it should stop processing that dataset, otherwise it tries to delete it.

The reason is the snapDestroy function. It tries to delete the snapshot, but since none is found, there is no @autorep- suffix. So it tries to delete the dataset entirely.

aaronhurt commented:
> But: when it skips a dataset that failed the “checkDataset “ function, it should stop processing that dataset, otherwise it tries to delete it.
> 
> The reason is the snapDestroy function. It tries to delete the snapshot, but since none is found, there is no @autorep- suffix. So it tries to delete the dataset entirely.
> 
> ```
> Aug 31 09:17:11 zfs-replicate.sh[78751]: cannot open 'tank/dst/1/src/1/2/3': dataset does not exist
> Aug 31 09:17:11 zfs-replicate.sh[78751]: attempting destroy tank/dst/1/src/1/2
> Aug 31 09:17:11 zfs-replicate.sh[78751]: failed - trying rename tank/dst/1/src/1/2 to tank/dst/1/src/1/recv-78927-1
> ```
> 
> It does fail though…
> 
> An if, for, continue, would probably solve it inside the createSnap function right when it does the checkDataset

I'm not seeing this same behavior. It skips the entire pair. Is that using the most recent code?

```
_testZFSReplicate/snapCreateWithHostCheckErrors
0 creating lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.snapshot.lock
1 checking dataset cmd=./zfs.sh list -H -o name srcPool0/srcFS0
2 srcPool0/srcFS0@autorep-test1
3 srcPool0/srcFS0@autorep-test2
4 srcPool0/srcFS0@autorep-test3
5 checking dataset cmd=./zfs.sh list -H -o name dstPool0/dstFS0
6 dstPool0/dstFS0@autorep-test1
7 dstPool0/dstFS0@autorep-test2
8 dstPool0/dstFS0@autorep-test3
9 found old snapshot srcPool0/srcFS0@autorep-test1
10 destroying snapshot cmd=./zfs.sh destroy srcPool0/srcFS0@autorep-test1
11 creating snapshot cmd=./zfs.sh snapshot srcPool0/srcFS0@autorep-08312024_1725121827
12 creating lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.send.lock
13 sending snapshot cmd=./zfs.sh send -Rs -I srcPool0/srcFS0@autorep-test3 srcPool0/srcFS0@autorep-08312024_1725121827 | ./zfs.sh receive -vFd dstPool0/dstFS0
14 receive -vFd dstPool0/dstFS0
15 deleting lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.send.lock
16 checking host cmd=false
17 WARNING: skipping replication set 'srcPool1/srcFS1/subFS1:dstPool1/dstFS1@dstHost1' - source or destination host check failed
18 checking host cmd=false
19 WARNING: skipping replication set 'srcPool2/srcFS2:dstPool2/dstFS2@dstHost2' - source or destination host check failed
20 checking host cmd=false
21 WARNING: skipping replication set 'srcPool3/srcFS3@srcHost3:dstPool3/dstFS3' - source or destination host check failed
22 checking host cmd=false
23 WARNING: skipping replication set 'srcPool4/srcFS4@srcHost4:dstPool4/dstFS4@dstHost4' - source or destination host check failed
24 deleting lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.snapshot.lock
```

tschettervictor commented:
> > But: when it skips a dataset that failed the “checkDataset “ function, it should stop processing that dataset, otherwise it tries to delete it.
> > The reason is the snapDestroy function. It tries to delete the snapshot, but since none is found, there is no @autorep- suffix. So it tries to delete the dataset entirely.
> > ```
> > Aug 31 09:17:11 zfs-replicate.sh[78751]: cannot open 'tank/dst/1/src/1/2/3': dataset does not exist
> > Aug 31 09:17:11 zfs-replicate.sh[78751]: attempting destroy tank/dst/1/src/1/2
> > Aug 31 09:17:11 zfs-replicate.sh[78751]: failed - trying rename tank/dst/1/src/1/2 to tank/dst/1/src/1/recv-78927-1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It does fail though…
> > An if, for, continue, would probably solve it inside the createSnap function right when it does the checkDataset
> 
> I'm not seeing this same behavior for the host check failure. Let me try to simulate a dataset failure.
> 
> ```
> _testZFSReplicate/snapCreateWithHostCheckErrors
> 0 creating lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.snapshot.lock
> 1 checking dataset cmd=./zfs.sh list -H -o name srcPool0/srcFS0
> 2 srcPool0/srcFS0@autorep-test1
> 3 srcPool0/srcFS0@autorep-test2
> 4 srcPool0/srcFS0@autorep-test3
> 5 checking dataset cmd=./zfs.sh list -H -o name dstPool0/dstFS0
> 6 dstPool0/dstFS0@autorep-test1
> 7 dstPool0/dstFS0@autorep-test2
> 8 dstPool0/dstFS0@autorep-test3
> 9 found old snapshot srcPool0/srcFS0@autorep-test1
> 10 destroying snapshot cmd=./zfs.sh destroy srcPool0/srcFS0@autorep-test1
> 11 creating snapshot cmd=./zfs.sh snapshot srcPool0/srcFS0@autorep-08312024_1725121827
> 12 creating lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.send.lock
> 13 sending snapshot cmd=./zfs.sh send -Rs -I srcPool0/srcFS0@autorep-test3 srcPool0/srcFS0@autorep-08312024_1725121827 | ./zfs.sh receive -vFd dstPool0/dstFS0
> 14 receive -vFd dstPool0/dstFS0
> 15 deleting lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.send.lock
> 16 checking host cmd=false
> 17 WARNING: skipping replication set 'srcPool1/srcFS1/subFS1:dstPool1/dstFS1@dstHost1' - source or destination host check failed
> 18 checking host cmd=false
> 19 WARNING: skipping replication set 'srcPool2/srcFS2:dstPool2/dstFS2@dstHost2' - source or destination host check failed
> 20 checking host cmd=false
> 21 WARNING: skipping replication set 'srcPool3/srcFS3@srcHost3:dstPool3/dstFS3' - source or destination host check failed
> 22 checking host cmd=false
> 23 WARNING: skipping replication set 'srcPool4/srcFS4@srcHost4:dstPool4/dstFS4@dstHost4' - source or destination host check failed
> 24 deleting lockfile /var/folders/pg/d8xk8vw5235172fkh8vk98g40000gn/T//.replicate.snapshot.lock
> ```

The dataset it fails at is the "destination with remote host" so the right side of the ":"

aaronhurt commented:
> The dataset it fails at is the "destination with remote host" so the right side of the ":"

It shouldn't matter, ssh should return non-zero if the remote command returns non-zero.

```
[ahurt@schroeder ~/zfs-replicate]$ zfs list -H -o name fake/fake; echo $?
cannot open 'fake/fake': dataset does not exist
1
```
[ahurt@charlie ~]$ ssh schroeder zfs list -H -o name fake/fake; echo $?
cannot open 'fake/fake': dataset does not exist
1
```

tschettervictor commented:
> > The dataset it fails at is the "destination with remote host" so the right side of the ":"
> 
> It shouldn't matter, ssh should return non-zero if the remote command returns non-zero.
> 
> ```
> [ahurt@schroeder ~/zfs-replicate]$ zfs list -H -o name fake/fake; echo $?
> cannot open 'fake/fake': dataset does not exist
> 1
> ```
> 
> [ahurt@charlie ~]$ ssh schroeder zfs list -H -o name fake/fake; echo $? cannot open 'fake/fake': dataset does not exist 1

See above. User error. 
But still, would there be a way to account for an error like that?

aaronhurt commented:
> See above. User error. But still, would there be a way to account for an error like that?
Let me see if I can mock up a test, and then handle it.

aaronhurt commented:
Looking at the main scripts,`zfs-replicate.sh` is 560 lines and `test/test.sh` is 384 lines. We're almost there 🤣

tschettervictor commented:
Here is the log. I first did a replication with RECURSE_CHILDREN=1, the tried doing the exact same replication with RECURSE_CHILDREN=0

```
Aug 31 12:12:52 zfs-replicate.sh[80977]: sending snapshot cmd=zfs send -Rs -I tank/src/1@autorep-08312024_1725127961 tank/src/1@autorep-08312024_1725127971 | zfs receive -vFd tank/dst/1
Aug 31 12:12:52 zfs-replicate.sh[80977]: WARNING: skipping dataset tank/src/1/2 and its children: snapshot autorep-08312024_1725127971 does not exist
Aug 31 12:12:54 zfs-replicate.sh[80977]: cannot open 'tank/dst/1/src/1/2/3@autorep-08312024_1725127957': dataset does not exist
Aug 31 12:12:54 zfs-replicate.sh[80977]: cannot open 'tank/dst/1/src/1/2/3@autorep-08312024_1725127961': dataset does not exist
Aug 31 12:12:54 zfs-replicate.sh[80977]: cannot open 'tank/dst/1/src/1/2/3': dataset does not exist
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1@autorep-08312024_1725127957
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/2@autorep-08312024_1725127957
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/2@autorep-08312024_1725127961
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/2
Aug 31 12:12:56 zfs-replicate.sh[80977]: failed - trying rename tank/dst/1/src/1/2 to tank/dst/1/src/1/recv-81180-1
Aug 31 12:12:56 zfs-replicate.sh[80977]: another pass:
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/recv-81180-1
Aug 31 12:12:56 zfs-replicate.sh[80977]: failed (0)
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/recv-81180-1/3@autorep-08312024_1725127957
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/recv-81180-1/3@autorep-08312024_1725127961
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/recv-81180-1/3
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: another pass:
Aug 31 12:12:56 zfs-replicate.sh[80977]: attempting destroy tank/dst/1/src/1/recv-81180-1
Aug 31 12:12:56 zfs-replicate.sh[80977]: success
Aug 31 12:12:56 zfs-replicate.sh[80977]: receiving incremental stream of tank/src/1@autorep-08312024_1725127971 into tank/dst/1/src/1@autorep-08312024_1725127971
Aug 31 12:12:57 zfs-replicate.sh[80977]: received 312B stream in 1 seconds (312B/sec)
Aug 31 12:12:57 zfs-replicate.sh[80977]: deleting lockfile /tmp/.replicate.send.lock
Aug 31 12:12:57 zfs-replicate.sh[80977]: deleting lockfile /tmp/.replicate.snapshot.lock
Aug 31 12:12:57 zfs-replicate.sh[80977]: pruning 2 logs
Aug 31 12:12:57 zfs-replicate.sh[80977]: /mnt/tank/logs/autorep-08312024_1725117429.log
Aug 31 12:12:57 zfs-replicate.sh[80977]: /mnt/tank/logs/autorep-08312024_1725110220.log
Aug 31 12:12:57 zfs-replicate.sh[80977]: success total sets 1 skipped 0

```

aaronhurt commented:
Hrm, yeah that's definitely odd but I see what it's trying to do. That's all zfs itself trying to reconcile state for the replication stream. The script isn't sending those `destroy` commands. The string `attempting` doesn't exist in the script. The part I find strange is that it logs a send and then immediately logs a skip. There should be other actions between those two pieces. The script is designed to run synchronously. That send and receive pipe should complete before the script tries to move on to another dataset. That doesn't make much sense. I added some simulated delay in the test receive command, and it is indeed sequential. The script blocks till the receive pipe completes.

tschettervictor commented:
Before it can complete the send, the destination has to match up. So it destroys the existing ones, then completes the send.

tschettervictor commented:
In any case I just switched back and forth between RECURSE_CHILDREN=1/0 and it seemed to work fine.

tschettervictor commented:
It skips because it is doing a non-recursive replication with nested datasets. It’s seeing the datasets, but also seeing it’s not supposed to do a recursive send. So it’s skips the nested ones (which are included in the send anyway, except as directories I suppose)
